FIX: allow colorbar mappable norm to change and do right thing

jklymak opened the pull request:
## PR Summary

Closes #13228 

```python

import matplotlib.pyplot as plt
import matplotlib as mpl
import numpy as np

N = 100
X, Y = np.mgrid[0:3:complex(0, N), 0:2:complex(0, N)]
Z1 = (1 + np.sin(Y * 10.)) * X**(2.)
data = Z1

fig, ax = plt.subplots()
cax = ax.pcolormesh(data, cmap='RdBu_r', rasterized=True)
cbar = fig.colorbar(cax, ax=ax)
cax.set_norm(mpl.colors.PowerNorm(gamma=2, vmin=0.001, vmax=18))

```

## Before

Note this colorbar is simply wrong...

![before](https://user-images.githubusercontent.com/1562854/51433940-7aaeba00-1c0b-11e9-90e2-90aa31359f6a.png)


## After
![new](https://user-images.githubusercontent.com/1562854/51433935-6074dc00-1c0b-11e9-97d3-9c1f6c54c6b0.png)

# Changing norm erases colorbar formatter and locator...

If the norm is changed, then any formatter and locator settings are also lost.  This is new behaviour, but I think reasonable if the user has changed the norm.  Here we illustrate this just by setting the ticks (which changes the Locator to FixedLocator).  Note that when the norm is set the locator goes back to the automatic locator.  

```python
fig, ax = plt.subplots()
cax = ax.pcolormesh(data, cmap='RdBu_r', rasterized=True)
cbar = fig.colorbar(cax, ax=ax)

print('original', cbar.locator)
cbar.set_ticks([0, 10, 20])
print('first set ticks', cbar.locator)
cax.set_norm(mpl.colors.PowerNorm(gamma=2, vmin=0.001, vmax=18))
print('after norm changed', cbar.locator)
cbar.set_ticks(np.arange(0, 20, 2))
print('after ticks set second time', cbar.locator)
plt.show()
```


```
original <matplotlib.colorbar._ColorbarAutoLocator object at 0x127311be0>
first set ticks <matplotlib.ticker.FixedLocator object at 0x127351320>
after norm changed <matplotlib.colorbar._ColorbarAutoLocator object at 0x127311d68>
after ticks set second time <matplotlib.ticker.FixedLocator object at 0x1272f1080>
```

# Next issue: Reset scale if norm changes...

The axis scale wasn't being reset if the norm changed, but it needs to be reset as well...

```python
print('changing norm', cbar.ax.yaxis.get_scale())
pcm.set_norm(mcolors.LogNorm(vmin=1, vmax=100))
print('norm changed', cbar.ax.yaxis.get_scale())
pcm.set_norm(mcolors.Normalize(vmin=-20, vmax=20))
print('norm changed', cbar.ax.yaxis.get_scale())
```

### Before
```
changing norm linear
norm changed linear
norm changed linear
```

### After
```
changing norm linear
norm changed log
norm changed linear
```

I'm marking WIP until @LevN0 lets us know we squashed all the bugs...

# Next Issue: format='%1.4e' now works...

See new test...

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is [Flake 8](http://flake8.pycqa.org/en/latest/) compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

<!--
Thank you so much for your PR!  To help us review your contribution, please
consider the following points:

- A development guide is available at https://matplotlib.org/devdocs/devel/index.html.

- Help with git and github is available at
  https://matplotlib.org/devel/gitwash/development_workflow.html.

- Do not create the PR out of master, but out of a separate branch.

- The PR title should summarize the changes, for example "Raise ValueError on
  non-numeric input to set_xlim".  Avoid non-descriptive titles such as
  "Addresses issue #8576".

- The summary should provide at least 1-2 sentences describing the pull request
  in detail (Why is this change required?  What problem does it solve?) and
  link to any relevant issues.

- If you are contributing fixes to docstrings, please pay attention to
  http://matplotlib.org/devel/documenting_mpl.html#formatting.  In particular,
  note the difference between using single backquotes, double backquotes, and
  asterisks in the markup.

We understand that PRs can sometimes be overwhelming, especially as the
reviews start coming in.  Please let us know if the reviews are unclear or
the recommended next step seems overly demanding, if you would like help in
addressing a reviewer's comments, or if you have been waiting too long to hear
back on your PR.
-->

jklymak commented:
Tests added.  Note that this obviates the need for the user to call `cbar.update_normal`, at least for the cases I know about.  ping @LevN0 and @efiring

NelleV commented:
This looks good! Thanks a lot for the detailed explanation. It made reviewing much faster.
(I'm not approving yet, as Tim's comment still stands)

jklymak commented:
OK, note added, though its a bit of a funky place to add a note about colorbar changes....

timhoffm commented on lib/matplotlib/cm.py:
```diff
@@ -349,6 +349,13 @@ def set_norm(self, norm):
+        Notes
+        -----
+        If there are any colorbars using the mappable for this norm, setting
+        the norm of the mappable will reeset the norm, locator, and formatters
```
Typo.
```suggestion
        the norm of the mappable will reset the norm, locator, and formatters
```

timhoffm commented:
> OK, note added, though its a bit of a funky place to add a note about colorbar changes....

True, but better than nothing and I don't have a better solution.

NelleV commented:
@jklymak I screwed up and didn't fix all the PEP8 from the github text editor… Do you mind fixing the rest (and maybe rebasing to squash the PEP8 fixes commits altogether?)

jklymak commented:
Thanks for the review @NelleV.  But this is still a WIP.  I mucked around with so much it needs a thorough cleaning now.   In particular for each colorbar draw_all is called three times which is unacceptable!

jklymak commented:
I think this is ready for now, but pinging @LevN0

Note that I decided we don't need to make locator and formatter properties, so I took that out to make things more straight forward...

timhoffm commented on lib/matplotlib/colorbar.py:
```diff
@@ -387,30 +387,27 @@ def __init__(self, ax, cmap=None,
         self.outline = None
         self.patch = None
         self.dividers = None
+        # self.locator and self.formatter are properties...
```
this comment is not true anymore, I think.

timhoffm commented on lib/matplotlib/colorbar.py:
```diff
@@ -449,14 +445,14 @@ def draw_all(self):
 
+    def set_norm(self, norm):
+        """
+        set the norm of the mappable associateed with this colorbar.
```
```suggestion
        Set the norm of the mappable associated with this colorbar.
```

timhoffm commented on lib/matplotlib/colorbar.py:
```diff
@@ -1105,6 +1131,7 @@ def on_mappable_changed(self, mappable):
         by :func:`colorbar_factory` and should not be called manually.
 
         """
+        _log.debug('mappable changed')
```
```suggestion
        _log.debug('colorbar mappable changed')
```
Giving slightly more context helps understanding the log output.

timhoffm commented on lib/matplotlib/colorbar.py:
```diff
@@ -1136,9 +1163,20 @@ def update_normal(self, mappable):
+        have been customized, they will need to be customized again.
         """
+
+        _log.debug('update normal')
```
```suggestion
        _log.debug('colorbar update normal')
```
Giving slightly more context helps understanding the log output.

timhoffm commented on lib/matplotlib/tests/test_colorbar.py:
```diff
@@ -424,23 +424,57 @@ def test_colorbar_renorm():
+    assert isinstance(cbar.locator, _ColorbarLogLocator)
+    assert np.allclose(cbar.ax.yaxis.get_majorticklocs(),
+                       np.logspace(-8, 5, 14))
+    # note that that set_norm reemoves the FixedLocator...
```
```suggestion
    # note that set_norm removes the FixedLocator
```

timhoffm commented on lib/matplotlib/tests/test_colorbar.py:
```diff
@@ -424,23 +424,57 @@ def test_colorbar_renorm():
 
 
+def test_colorbar_format():
+    # make sure that format is passed properly....
```
```suggestion
    # make sure that format is passed properly
```

jklymak commented:
Sorry for the WIP embargo and confusion that caused. 

Most of the issues identified in #13228 have been addressed and this is ready for a second review.

NelleV commented:
Is there a way to directly merge @timhoffm 's suggestion?

jklymak commented:
Thats strange, I thought I'd changed all the lines he asked for...  pushed and rebased a fix

jklymak commented:
(In general, yes, you can commit suggestions using the github interface, but sometimes its easier for the original PR writer to do it)

LevN0 commented:
Was just trying to figure out how to do a PR (in general), and had the question while reading the guidelines: should the update_normal change to reset locator/formatter go into doc/api/api_changes.rst?

efiring commented on lib/matplotlib/colorbar.py:
```diff
@@ -449,14 +444,14 @@ def draw_all(self):
         if self.filled:
             self._add_solids(X, Y, C)
 
+    def set_norm(self, norm):
```
I don't think that we want to have the colorbar in any way controlling the mappable.

jklymak replied:
OK, I somewhat agree.  OTOH, right now, the colorbar is inexplicably subclassed from `ScalarMappable` which is where `set_norm` comes from.  I have no idea why that is or how to extricate it, but I don't think we should have a `colorbar.set_norm` that does nothing.

jklymak replied:
... so actually making it *not* be a subclass of `ScalarMappable` *almost* works.  Is there a real reason this is a problem?

efiring replied:
I think the subclassing was just an unfortunate design choice on my part, long ago.  I suspect my thought was that it should be possible to make a colorbar without any accompanying mappable, so that was the starting point.  But I'm sure that can still be done in a straightforward way without subclassing.

jklymak commented:
@efiring I think we don't need colorbar to inherit from ScalarMappable, so I've taken that out and make deprectaed dummy functions for some ScalarMappable methods that folks may have been calling (and likely were not actually doing anything useful).

jklymak commented:
One annoying consequence of this PR is that *any* changes to the mappable wipe-out formatting, at least the way it is now.  I'm going back and forth on this whether its reasonable:

If we use properties, we can probably manage this by allowing the setter to toggle a boolean.  

```python
import matplotlib.pyplot as plt
import matplotlib as mpl
import matplotlib.colors as mcolors
import numpy as np

N = 100
X, Y = np.mgrid[0:3:complex(0, N), 0:2:complex(0, N)]
Z1 = (1 + np.sin(Y * 10.)) * X**(2.)
data = Z1

fig, ax = plt.subplots()
pcm = ax.imshow(data, cmap='RdBu_r', rasterized=True)
cbar = fig.colorbar(pcm, ax=ax, format='%4.2e')
fig.savefig('BeforeRelim.png')

pcm.set_clim(-10, 10)
fig.savefig('AfterRelim.png')
plt.show()
```

So, the second plot loses the result of the `format` kwarg.

### Before Relim:

![beforerelim](https://user-images.githubusercontent.com/1562854/51504708-3ea86000-1d97-11e9-928c-2ee7919ba2c0.png)


### After Relim:
![afterrelim](https://user-images.githubusercontent.com/1562854/51504700-305a4400-1d97-11e9-9962-1f2020946157.png)

jklymak commented:
OK, so I have a version (using properties) that makes the user-set formatter or ticker persistent, including passing the `format='%f'` or `ticks=[1, 2, 3]` kwargs to `colorbar`.  That can lead to unexpected results when the norm is changed, so its a question of the lesser of two evils.  

Thumbs up: Make user hard-coded ticks, locators, formatters, etc persistent across mappable norm changes

Thumbs down: if the norm changes (including `mappable.set_clim`; see above), the user-set formatting gets reset and the default is used.  

Note that for an axis if we change the `scale`, which is the same as chaning the `norm` in this case, the user's ticks are lost.  OTOH, if we change the limits without changing the scale (which is also what a norm changes) the ticks are *not* lost.

timhoffm commented:
> so I've taken that [ScalarMappable subclassing] out and make deprectaed dummy functions for some ScalarMappable methods that folks may have been calling

You could take all that stuff of a `ColorbarMappableMixin`. That makes it more clear which parts of `Colorbar` are just back-compat, which makes it easier to remove them later. Also, since they are in a base class, these methods will no show up in the `Colorbar` docs (same as the ScalarMapple functions do not show up currently).

jklymak commented:
That seems reasonable   I assume it would be private?

jklymak commented:
I think this is ready again:

 - fixed it so that changes to the norm vmin/vmax don't affect the formatter or locator, but actual changes to the norm do change the formatter/locator.  Note modified test checks this behaviour.  
 - no longer made ColorbarBase a subclass of ScalarMappable.
 - moved the deprecated methods that used to be from ScalarMappable into their own private class (mixin?)

jklymak commented:
ping @efiring  (@NelleV your review may be stale as well).

efiring commented on lib/matplotlib/colorbar.py:
```diff
@@ -304,7 +304,47 @@ def tick_values(self, vmin, vmax):
+    Private class to hold deprecated ColorbarBase methods that used to be
+    inhereted from ScalarMappable.
+    """
+    @cbook.deprecated("3.1", alternative="mappable.set_norm")
```
For alternative, did you mean `ScalarMappable.set_norm` here, as below?

jklymak replied:
Fixed both points!  Thanks,

jklymak commented:
Did anyone else have time to look at this again?  Changes from previous approvals are [here](
https://github.com/matplotlib/matplotlib/pull/13234#issuecomment-457027060)

LevN0 commented:
Probably not addressed to me, but I looked at it again to confirm it still addresses the original issue, and it does.

NelleV commented on lib/matplotlib/colorbar.py:
```diff
@@ -304,7 +304,47 @@ def tick_values(self, vmin, vmax):
+
+    @cbook.deprecated("3.1", alternative="ScalarMappable.get_cmap")
+    def get_cmap(self):
+        'return the colormap'
```
Minor comment: this should be a docstring, even if it is a deprecated method.

NelleV commented on lib/matplotlib/colorbar.py:
```diff
@@ -304,7 +304,47 @@ def tick_values(self, vmin, vmax):
+
+    @cbook.deprecated("3.1", alternative="ScalarMappable.get_clim")
+    def get_clim(self):
+        'return the min, max of the color limits for image scaling'
```
Same here

jklymak replied:
I changed to 
```python
"""
return the...
"""
```

Is that all you meant, or did you want a whole doctoring?  Its a private deprecated class, so it won't show up in the docs...

jklymak replied:
... also these *were* just directly copied from `cm.ScalarMappable`.

NelleV replied:
Yeah, I've seen things like this before on Matplotlib, but strings just hanging around are not super good practice.

NelleV commented:
Thanks!
